### ISSUE TITLE
improve caching on prow

### ISSUE THREAD

**Issue by BenTheElder** [kubernetes/test-infra#5436]
Right now jobs building with bazel use a cache path on an SSD on the node, which is restricted to one job at a time by reserving a fixed port on the node. We can improve the concurrency of this with:

- jobs using the bazel cache should use one port
- docker-in-docker jobs using the docker graph cache should use another
- jobs in different repos should use different cache dirs for the two above
- jobs targeting specific branches should probably use a different cache from other jobs in the same repo
--
- each of these cache directories needs to be mapped to a reserved node port so we don't accidentally use them concurrently.

/area prow
/area jobs
/assign
/cc @krzyzacy @cjwagner

**Comment by BenTheElder** [comment]
This should improve job throughput, and is becoming more of a problem now that we have many more jobs on Prow and are adding more repos.

**Comment by spxtr** [comment]
It looks like pod antiaffinity is beta now, we should use that.

**Comment by BenTheElder** [comment]
we definitely need to look into https://github.com/kubernetes/test-infra/issues/5449 first

**Comment by BenTheElder** [comment]
https://kubernetes.io/docs/concepts/configuration/assign-pod-node/#inter-pod-affinity-and-anti-affinity-beta-feature
Discussed this with @krzyzacy today, we think pod anti-affinity labels + per-repo cache paths on the host   should be a good start.

**Comment by BenTheElder** [comment]
We really just need better caching mechanisms, we can semi-trivially pack more jobs onto larger nodes, but we can only get quota so many local SSDs for caching.
  
What we need to do is remove having the node be exclusively held by every presubmit until it completes in order to use the cache SSD at all, the majority of jobs waiting to schedule are just waiting on `NodeHasHostPorts` while the cluster has plenty of spare resources (I'm seeing 20, 30 minutes just waiting for a host port while the cluster has loads of CPU/RAM!). There are a couple of routes for this:

- Use shared builds to make more jobs not actually check out the repo

- Remove the need for an exclusive git cache:
  - [This option was used before and reverted](https://github.com/kubernetes/kubernetes/pull/24822), but I looked again and won't work for us: Using shallow clones and no actual caching. 
    - This won't work because lots of things depend on the `git describe`, which I've not been able to get working without essentially fetching the whole repo anyhow.
  - `rsync`ing / `cp`ing a read-only copy updated by some other job. This *could* work but we'd still need to fetch updates and somehow we have to safely update the actual cache...
  - Similar to above we can use git's built in support to use a cache with: [`git clone --reference-if-able /path/to/git/cache --dissociate ...`](https://git-scm.com/docs/git-clone#git-clone---reference-if-ableltrepositorygt).
    - It appears to be safe to update the reference while cloning referencing it, since sha'd objects are referenced and git will fall back on fetching them over the network, and dissociate will repack them to not depend on the reference. 
    - We can have a job assigned to updating the cache or periodically do this as a maintenance task ourselves, the problem with not having a cache seems to be fetching the *entire* history of k/k, not the recent portions of it. 
    - I've been researching this more deeply and I think this is safe enough to start canary testing a version of it on Monday. 

- Generally seperate out the different kinds of caching. A CI job using docker-in-docker or bazel doesn't need to have the git cache as well, restrict these cache directories per repo per node separately.

- Stop pretending to use the cache when we aren't. If a job is using bazel but is using `bazel clean --expunge` we shouldn't even point it at a cache then. We're needlessly limiting our throughput here, particularly when considering smaller repos like test-infra which definitely don't need to peg a whole node.

**Comment by krzyzacy** [comment]
(let's see how'd 🔥  will be after code freeze)

**Comment by BenTheElder** [comment]
> (let's see how'd 🔥 will be after code freeze)

We don't really have to wait for code freeze to canary test the read-only parts, and [the code freeze ends tomorrow anyhow](https://github.com/kubernetes/features/blob/master/release-1.9/release-1.9.md#timeline). We should at least begin exploring:
- split docker data-root per-repo
- split ports / affinities to match
- try canary testing `git clone --reference /path/to/cache --dissociate ...`, if this works fine (it should?) then we can stop restricting the git cache
- move bazel to empty dir when we aren't actually using the cache, also ideally per repo / per release when we are using the cache

We have plenty of space to play with on those SSDs and we can do better than "one thing at a time may use our 375gb disk"

**Comment by BenTheElder** [comment]
shared-builds are looking pretty good tentatively, thanks @cjwagner for fixing the statuses and @luxas for letting me steal a copy of kubeadm to kanary and for continuing to use shared-build even when we weren't convinced it was going to work out :-) !
see: https://github.com/kubernetes/test-infra/pull/5832#issuecomment-351230605

**Comment by stevekuznetsov** [comment]
I was messing around also with using a main `git` checkout on the SSD volume and `git submodule` to disseminate checkouts to other volumes that tests would use -- seemed like one possible option for concurrency at least on the checkout step.

**Comment by BenTheElder** [comment]
> I was messing around also with using a main git checkout on the SSD volume and git submodule to disseminate checkouts to other volumes that tests would use -- seemed like one possible option for concurrency at least on the checkout step.

Yeah, another option specifically for git is to use an init container and just do a plain update of the cached clone and then copy it to a workspace for the job in the init. It looks like we might be able to request a port in the init container and then release our hold on the node when it exits.

I think we still want to move towards an init container for checkout anyhow, but I also like the `git clone --reference --dissociate` option because it looks like we could actually have multiple ongoing checkouts on the same machine at the same time. I'm not sure about the submodule route.

**Comment by BenTheElder** [comment]
we've kicked a lot of jobs out of using the hostport junk, and when we have the checkout container we can just put k8s in that. for build / test caching we use bazel remote caching now instead of caching on the nodes.

superseded by other issues
/close
